File: Leet_Code-1249/Leet_Code-1249.py
'''
Question 1249
'''

class Solution:
    def minRemoveToMakeValid(self, s: str) -> str:
        c_open = []
        c_close = []
        i = 0
        n = len(s)
        while i < n:
            if s[i] == '(':
                c_open.append(i)
            elif s[i] == ')':
                c_close.append(i)
                if len(c_close) > len(c_open):
                    s = s[:i]+s[i+1:]
                    c_close.pop()
                    n -=1
                    i -= 1
            i += 1
        while len(c_open) > len(c_close):
            c = c_open[-1]
            s = s[:c]+s[c+1:]
            c_open.pop()
        return s

File: Leet_Code-1249/Leet_Code-1249.java
/*
 * Question 1249
 */

import java.util.ArrayList;
class Solution {
    public String minRemoveToMakeValid(String s) {
        ArrayList<Integer> c_open = new ArrayList();
        ArrayList<Integer> c_close = new ArrayList();
        int i = 0;
        int n = s.length();
        while (i < n) {
            if (s.charAt(i) == '(') {
                c_open.add(i);
            } else if (s.charAt(i) == ')') {
                c_close.add(i);
                if (c_close.size() > c_open.size()) {
                    s = s.substring(0,i)+s.substring(i+1,s.length());
                    c_close.remove(c_close.size()-1);
                    n--;
                    i--;
                };
            };
            i++;
        };
        while (c_open.size() > c_close.size()) {
            int c = c_open.get(c_open.size()-1);
            s = s.substring(0,c)+s.substring(c+1,s.length());
            c_open.remove(c_open.size()-1);
        };
        return s;
    }
}